File: java/Medium/_95_UniqueBinarySearchTreesII.java
package Medium;

import java.util.*;

import org.testng.annotations.Test;

import CS.TreeNode;

public class _95_UniqueBinarySearchTreesII {
    public List<TreeNode> generateTrees(int n) {
        return helper(1, n);
    }

    private List<TreeNode> helper(int low, int high){
        List<TreeNode> res = new ArrayList<>();
        if(high<low){
            res.add(null);
            return res;
        } 

        for(int root = low; root<=high; root++){
            List<TreeNode> left = helper(low, root-1);
            List<TreeNode> right = helper(root+1, high);

            for(TreeNode l : left){
                for(TreeNode r : right){
                    res.add(new TreeNode(root, l, r));
                }
            }
        }

        return res;
    }

   

    @Test
    public void test(){
        List<TreeNode> test = generateTrees(3);
    }
}

File: Python3/_95_UniqueBinarySearchTreesII.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[TreeNode]:
        
        return self.helper(1,n)

    def helper(self, low:int, high:int) -> List[TreeNode]:
        res = []
        if low > high:
            res.append(None)
            return res

        for i in range(low, high+1):
            left = self.helper(low, i-1)
            right = self.helper(i+1, high)

            for l in left:
                for r in right:
                    res.append(TreeNode(i,l,r))
        return res

test = Solution()
test.generateTrees(3)